identity/service_principals.go

```diff
@@ -0,0 +1,300 @@
+		return err
+	}
+	updateRequest := ru.toRequest()
+	updateRequest.Groups = servicePrincipal.Groups
+	updateRequest.Roles = servicePrincipal.Roles
+	return a.C.Scim(http.MethodPut,
+		fmt.Sprintf("/preview/scim/v2/ServicePrincipals/%v", servicePrincipalID),
+		updateRequest, nil)
+}
+
+// PatchR updates resource-friendly entity
+func (a ServicePrincipalsAPI) PatchR(servicePrincipalID string, r patchRequest) error {
+	return a.C.Scim(http.MethodPatch, fmt.Sprintf("/preview/scim/v2/ServicePrincipals/%v", servicePrincipalID), r, nil)
+}
+
+// Update will update the servicePrincipal given the servicePrincipal id, application id, display name, entitlements and roles
+func (a ServicePrincipalsAPI) Update(servicePrincipalID string, applicationId string, displayName string, entitlements []string, roles []string) error {
+	servicePrincipalPath := fmt.Sprintf("/preview/scim/v2/ServicePrincipals/%v", servicePrincipalID)
+	scimServicePrincipalUpdateRequest := struct {
+		Schemas       []URN                  `json:"schemas,omitempty"`
+		ApplicationId string                 `json:"applicationId,omitempty"`
+		Entitlements  []EntitlementsListItem `json:"entitlements,omitempty"`
+		DisplayName   string                 `json:"displayName,omitempty"`
+		Roles         []RoleListItem         `json:"roles,omitempty"`
+		Groups        []GroupsListItem       `json:"groups,omitempty"`
+	}{}
+	scimServicePrincipalUpdateRequest.Schemas = []URN{ServicePrincipalSchema}
+	scimServicePrincipalUpdateRequest.ApplicationId = applicationId
+	scimServicePrincipalUpdateRequest.DisplayName = displayName
+	scimServicePrincipalUpdateRequest.Entitlements = []EntitlementsListItem{}
+	for _, entitlement := range entitlements {
+		scimServicePrincipalUpdateRequest.Entitlements = append(scimServicePrincipalUpdateRequest.Entitlements, EntitlementsListItem{Value: Entitlement(entitlement)})
+	}
+	scimServicePrincipalUpdateRequest.Roles = []RoleListItem{}
+	for _, role := range roles {
+		scimServicePrincipalUpdateRequest.Roles = append(scimServicePrincipalUpdateRequest.Roles, RoleListItem{Value: role})
+	}
+	//Get any existing groups that the servicePrincipal is part of
+	servicePrincipal, err := a.read(servicePrincipalID)
+	if err != nil {
+		return err
+	}
+	scimServicePrincipalUpdateRequest.Groups = servicePrincipal.Groups
+	return a.C.Scim(http.MethodPut, servicePrincipalPath, scimServicePrincipalUpdateRequest, nil)
+}
+
+// Delete will delete the servicePrincipal given the servicePrincipal id
+func (a ServicePrincipalsAPI) Delete(servicePrincipalID string) error {
+	servicePrincipalPath := fmt.Sprintf("/preview/scim/v2/ServicePrincipals/%v", servicePrincipalID)
+	return a.C.Scim(http.MethodDelete, servicePrincipalPath, nil, nil)
+}
+
+// SetServicePrincipalAsAdmin will add the servicePrincipal to a admin group given the admin group id and servicePrincipal id
+func (a ServicePrincipalsAPI) SetServicePrincipalAsAdmin(servicePrincipalID string, adminGroupID string) error {
```

Comment:
remove. this is definitely copy-paste from deprecated code

Reply:
2c313ea0d30a96f5ec12d5aa1572c2fc02c32bd9